cmd/clusterctl/client/cluster/topology.go

```diff
@@ -0,0 +1,641 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cluster
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	jsonpatch "github.com/evanphx/json-patch"
+	"github.com/gobuffalo/flect"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/component-base/featuregate"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	crwebhook "sigs.k8s.io/controller-runtime/pkg/webhook"
+
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/client/cluster/internal/dryrun"
+	"sigs.k8s.io/cluster-api/feature"
+	clusterctrl "sigs.k8s.io/cluster-api/internal/controllers/topology/cluster"
+	"sigs.k8s.io/cluster-api/internal/webhooks"
+)
+
+const (
+	maxClusterPerInput        = 1
+	maxClusterClassesPerInput = 1
+)
+
+// TopologyClient has methods to work with ClusterClass and ManagedTopologies.
+type TopologyClient interface {
+	DryRun(in *DryRunInput) (*DryRunOutput, error)
+}
+
+// topologyClient implements TopologyClient.
+type topologyClient struct {
+	proxy           Proxy
+	inventoryClient InventoryClient
+}
+
+// ensure topologyClient implements TopologyClient.
+var _ TopologyClient = &topologyClient{}
+
+// newTopologyClient returns a TopologyClient.
+func newTopologyClient(proxy Proxy, inventoryClient InventoryClient) TopologyClient {
+	return &topologyClient{
+		proxy:           proxy,
+		inventoryClient: inventoryClient,
+	}
+}
+
+// DryRunInput defines the input for DryRun function.
+type DryRunInput struct {
+	Objs              []*unstructured.Unstructured
+	TargetClusterName string
+}
+
+// PatchSummary defined the patch observed on an object.
+type PatchSummary = dryrun.PatchSummary
+
+// ChangeSummary defines all the changes detected by the Dryrun execution.
+type ChangeSummary = dryrun.ChangeSummary
+
+// DryRunOutput defines the output of the DryRun function.
+type DryRunOutput struct {
+	Clusters          []client.ObjectKey
+	ClusterClasses    []client.ObjectKey
+	ReconciledCluster client.ObjectKey
+	*ChangeSummary
+}
+
+func (t *topologyClient) DryRun(in *DryRunInput) (*DryRunOutput, error) {
```

Comment:
@fabriziopandini

Reply:
I would say no, at least for consistency with all the other public methods.
If we should introduce context, let's do it across the board with a single braking change, but I would first make some additional considerations on why we need it/where it will be used inside those methods